documenters_aggregator/spiders/ward_night.py

```diff
@@ -0,0 +1,278 @@
+        """
+        Yields a dictionary with the required keys.
+        """
+
+        rows = json.loads(response.body.decode('utf-8'))['values']
+
+        for row in rows:
+            # Strip leading or trailing whitespace from all values
+            for i in range(len(row)):
+                row[i] = row[i].strip()
+
+            # The JSON omits values for trailing columns with no values. By padding
+            # the rows out, the rest of the code can assume there will always be 14
+            # columns.
+            missing_values = 14 - len(row)
+            row.extend([''] * missing_values)
+            for result in self._parse_row(row):
+                yield result
+
+    def _days_for_frequency(self, frequency, day_of_week):
+        calendar = Calendar(self.start_date)
+        occurance_map = ['Monthly (1st occurrence)', 'Monthly (2nd occurrence)',
+                         'Monthly (3rd occurrence)', 'Monthly (4th occurrence)']
+
+        if frequency in occurance_map:
+            n = occurance_map.index(frequency) + 1
+            return calendar.nth_weekday(n, day_of_week)
+        elif frequency == 'Weekly':
+            return calendar.weekday(day_of_week)
+        elif frequency == 'Monthly (last occurrence)':
+            return calendar.last_weekday(day_of_week)
+        else:
+            # also handles 'Irregularly'
+            return []
+
+    def _parse_row(self, row):
+
+        if row[Row.HAS_WARD_NIGHTS] != 'Yes':
+            return []
+
+        try:
+            assert len(row[Row.START_TIME]) > 0, 'start time must have a value'
+            assert len(row[Row.END_TIME]) > 0, 'end time must have a value'
+        except:
+            # TODO: replace this with the correct logging
+            print('bad data!')
+            print(row)
+            return []
+
+        def build_event(day):
+            dates = self._parse_date_time(row, day)
+            return {
+                '_type': 'event',
+                'id': self._parse_id(row, dates['datetime']),
+                'name': self._parse_name(row),
+                'description': self._parse_description(row),
+                'classification': self._parse_classification(row),
+                'start_time': dates['start'],
+                'end_time': dates['end'],
+                'all_day': self._parse_all_day(row),
+                'status': self._parse_status(row),
+                'location': self._parse_location(row),
+            }
+
+        days = self._days_for_frequency(row[Row.FREQUENCY], row[Row.DAY_OF_WEEK])
+        return [build_event(day) for day in days]
+
+    def _parse_id(self, row, date):
+        """
+        Generate ID. We are assuming that there will only be a single event in each
+        ward in a single day.
+        """
+
+        values = {
+            'ward': row[Row.WARD],
+            'date': date.strftime('%Y-%m-%d')
+        }
+        id = 'ward{ward}-{date}'.format(**values)
+        return id
```

Comment:
Are these the standardized ids?

Reply:
No, they aren't! This predates that PR, but I will make that change. Good catch.